File: kata1/kata.py

# def make_kata(input_string):
#     """Test finished in 2.167048931121826 seconds """
#     input_string = input_string.lower()
#     output = ''
#     inputted_list = list(input_string)
#     inputted_set = set(inputted_list)
#     for symbol in inputted_set:
#         inputted_list.remove(symbol)
#     for symbol in input_string:
#         output += ')' if symbol in inputted_list else '('
#     return output

# def make_kata(inputted):
#     """ Test finished in 0.9294266700744629 seconds """
#     d = {}
#     inputted = inputted.lower()
#     outputted = ''
#     for symbol in inputted:
#         d[symbol] = ')' if symbol in d else '('
#     for symbol in inputted:
#         outputted += d[symbol]
#     return outputted

# def make_kata(inputted):
#     d = {}
#     def f(stri):
#         length = len(stri)
#         if length == 2:
#             stri = stri.lover()
#             if
#             if stri[0] == stri[1]:
#                 d[stri[0]] = ')'

#     d = set()
#     outputted = ''
#     outputted_list = []
#     for symbol in inputted:
#         symbol = symbol.lower()
#         outputted_list.append() = ')' if symbol in d else '('
#     for symbol in inputted:
#         outputted += d[symbol]
#     return outputted


# def make_kata(inputted):
#     """Test finished in 2.322965145111084 seconds"""
#     d = {}
#     inputted = inputted.lower()
#     inputted_list = list(inputted)
#     while inputted_list:
#         symbol = inputted_list.pop(0)
#         if symbol not in d:
#             d[symbol] = ')' if symbol in inputted_list else '('
#     for symbol in d:
#         inputted = inputted.replace(symbol, d[symbol])
#     return inputted

# def make_kata(inputted):
#     """Test finished in 2.30079984664917 seconds"""
#     d = {}
#     inputted = inputted.lower()
#     inputted_list = list(inputted)
#     while inputted_list:
#         symbol = inputted_list.pop(0)
#         if symbol in d:
#             continue
#         d[symbol] = ')' if symbol in inputted_list else '('
#     for symbol in d:
#         inputted = inputted.replace(symbol, d[symbol])
#     return inputted

# def make_kata(inputted):
#     """Test finished in 1.7508909702301025 seconds"""
#     d = set()
#     inputted = inputted.lower()
#     while inputted[0] not in ('(', ')'):
#         symbol = inputted[0]
#         inputted = inputted[1:]
#         if symbol in d:
#             inputted += ')'
#             continue
#         if symbol in inputted:
#             d.add(symbol)
#             inputted += ')'
#             continue
#         inputted += '('
#     return inputted

# def make_kata(inputted):
#     """Test finished in 1.748342514038086 seconds"""
#     d = ''
#     inputted = inputted.lower()
#     while inputted[0] not in ('(', ')'):
#         symbol = inputted[0]
#         inputted = inputted[1:]
#         if symbol in d:
#             inputted += ')'
#             continue
#         if symbol in inputted:
#             d += symbol
#             inputted += ')'
#             continue
#         inputted += '('
#     return inputted

# def make_kata(inputted):
#     """Test finished in 1.7696144580841064 seconds"""
#     d = ''
#     inputted = inputted.lower()
#     symbol = inputted[0]
#     while  symbol not in ('(', ')'):
#         inputted = inputted[1:]
#         if symbol in inputted:
#             d += symbol
#             inputted += ')'
#             symbol = inputted[0]
#             continue
#         if symbol in d:
#             inputted += ')'
#             symbol = inputted[0]
#             continue
#         inputted += '('
#         symbol = inputted[0]
#     return inputted


def make_kata(inputted):
    """Test finished in 2.796117067337036 seconds"""
    inputted = inputted.lower()
    inputted_list = list(inputted)
    def gen(li):
        nonuniq = []
        while li:
            symbol = li.pop(0)
            if symbol in nonuniq:
                out = ')'
            elif symbol in li:
                out = ')'
                nonuniq.append(symbol)                
            else:
                out = '('
            yield out
    return ''.join(_ for _ in gen(inputted_list))

File: kata1/golang/kata.go
package kata

import "strings"

// import "fmt"

// func MakeKata(inputted string) string {
//     lower := strings.ToLower(inputted)

//     outputted := ""
//     d := make(map[string]string)

//     for _, symbol := range lower {
//         symbol := string(symbol)
//         if d[symbol] == "" {
//             d[symbol] = "("
//         } else {
//             d[symbol] = ")"
//         }
//     }
//     for _, symbol := range lower {
//         symbol := string(symbol)
//         outputted += d[symbol]
//     }
//     return outputted
// }

// func MakeKata(inputted string) string {
//     lower := strings.ToLower(inputted)
//     outputted := ""
//     d := make(map[rune]rune)

//     for _, symbol := range lower {
//         if d[symbol] == 0 {
//             d[symbol] = '('
//         } else {
//             d[symbol] = ')'
//         }
//     }
//     for _, symbol := range lower {
//         outputted += string(d[symbol])
//     }
//     return outputted
// }

// func MakeKata(input string) string {
//     input = strings.ToLower(input)
//     var out []rune = []rune{}
//     symbols := make(map[rune]rune)
//     for _, sym := range input {
//         switch symbols[sym] {
//         case ')':
//             continue
//         case '(':
//             symbols[sym] = ')'
//         default:
//             symbols[sym] = '('
//         }
//     }
//     for _, sym := range input {
//         out = append(out, symbols[sym])
//     }
//     return string(out)
// }

/*
func MakeKata(inp string) string {
    out := []rune(strings.ToLower(inp))
    dict := make(map[rune]rune)

    for _, symbol := range out {
        switch dict[symbol] {
        case ')':
            continue
        case '(':
            dict[symbol] = ')'
        default:
            dict[symbol] = '('

        }

    }
    for pos, symbol := range out {
        out[pos] = dict[symbol]
    }
    return string(out)
}
*/

func MakeKata(input string) string {
    output := []rune(strings.ToLower(input))
    d := make(map[rune]rune)
    for _, symbol := range output {
        switch d[symbol] {
        case ')':
            continue
        case '(':
            d[symbol] = ')'
        default:
            d[symbol] = '('
        }
    }
    for pos, symbol := range output {
        // if symbol == ')' && d[symbol] == ')' {
        //     continue
        // }
        output[pos] = d[symbol]
    }
    return string(output)

}
